### ISSUE TITLE
Regular expressions don't worki in setlocal, counter to the man page

### ISSUE THREAD

**Issue by manujchandra** [ranger/ranger#2355]
I have similar problem to [this](https://github.com/ranger/ranger/issues/443).

I have spaces in my path:

The following 2 lines work, but they permanently set the options for all sub-folders which is undesirable:
`setlocal path='/home/manuj/Productivity_System/04 BIN/04 RA RP SP/002 RELIEVING' sort mtime`
`setlocal path='/home/manuj/Productivity_System/04 BIN/04 RA RP SP/002 RELIEVING' sort_reverse true`

I tried moving $ within quotes but that did not work as shown in man page. Tried with ~ and " also did not work.

The following also do not work:

`setlocal path='/home/manuj/Productivity_System/04 BIN/04 RA RP SP/002 RELIEVING$' sort mtime`
`setlocal path='/home/manuj/Productivity_System/04 BIN/04 RA RP SP/002 RELIEVING$' sort_reverse true`

`setlocal path="~/Productivity_System/04 BIN/04 RA RP SP/002 RELIEVING$" sort mtime`
`setlocal path="~/Productivity_System/04 BIN/04 RA RP SP/002 RELIEVING$" sort_reverse true`

In other words, the $ is not being considered

ranger version: ranger 1.9.3
Python version: 3.8.5 (default, May 27 2021, 13:30:53) [GCC 9.3.0]
Locale: en_US.UTF-8

From this [page](https://www.systutorials.com/docs/linux/man/1-ranger/):

`path can be quoted with either single or double quotes to prevent unwanted splitting. path='~/dl dl$' or path=``~/dl dl$''`

But this is not working for me.

**Comment by toonn** [comment]
Does it work for you even when unquoted? I can't get the `$` to work in any instance.

**Comment by manujchandra** [comment]
I am afraid 

`setlocal path="~/Downloads$" sort mtime`
`setlocal path="~/Downloads$" sort_reverse true`

is also not working

**Comment by toonn** [comment]
I suspect it doesn't handle regular expressions at all. Thanks for the report!

**Comment by manujchandra** [comment]
Mention not. Ranger is a great tool. I am keen on making it my default file explorer.

**Comment by Euren** [comment]
Just got confused by this myself. The code is there to handle regular expressions but, in container/settings.py, the `_raw_set` function does `regex = re.compile(re.escape(path))` which escapes the $, turning it into `\$`. Seems there is some ongoing work in https://github.com/ranger/ranger/pull/1680 to address this.

Removing that escape *does* allow $ to make the setlocal setting non-recursive, but I don't know what other effects its removal may have. For instance, the PR mentions directories with special regex characters might cause issues.

https://github.com/ranger/ranger/blob/7e6a0aa6a144f6733a8000a2822c107c93e60c1a/ranger/container/settings.py#L279-L290

**Comment by toonn** [comment]
I've picked this PR up again. `setlocal` will keep the non-regex behavior, `setinpath` will be the same command, `setinregex` will make the old behavior available.

**Comment by Euren** [comment]
Sounds like a good plan. Thanks for the update!

**Comment by manujchandra** [comment]
Hi,

Sorry, just to confirm, will this work?:

```
setinregex path='~/Productivity_System/04 BIN/04 RA RP SP$' sort mtime
setinregex path='~/Productivity_System/04 BIN/04 RA RP SP$' sort_reverse true
```

Because I tried it and its not working.

**Comment by Euren** [comment]
For regex, use re= instead of path=

**Comment by manujchandra** [comment]
Hi, sorry for the noob question:

```
setinregex re='~/Productivity_System/04 BIN/04 RA RP SP$' sort mtime
setinregex re='~/Productivity_System/04 BIN/04 RA RP SP$' sort_reverse true
```

Does not work either. Do I have to make a new build of Ranger or try in my existing installation?

Thanks.

**Comment by Euren** [comment]
New build as you need to somehow pull in the latest changes. Not sure how you currently have it installed though.
For a quick test, you can clone the repo and run ranger.py.

**Comment by manujchandra** [comment]
Got it. I will do the testing over the weekend. If anything else needs to be tested let me know. I will try to do them at the same time.

**Comment by toonn** [comment]
@manujchandra, could you make a screenrecording of this? Sounds like a weird issue.

**Comment by manujchandra** [comment]
Video is attached below.

[video.zip](https://github.com/ranger/ranger/files/7149217/video.zip)

**Comment by toonn** [comment]
@manujchandra, could you test it outside of Xonsh, just to make sure that's not the problem, since it's not a very common shell, bash, zsh and fish are all pretty popular among ranger users.

If that makes no difference you'll have to bisect, in the  repo run the following (full explanation as I don't know how familiar you are with git).

```
> git bisect start master v1.9.3
# Git will automatically check out a revision about midway between v1.9.3
# and the current tip of the branch.
> ./ranger.py # Verify whether or not it works.
> git bisect bad # If it didn't work
> git bisect good # If it did work
# Then Git'll check out a new revision midway between the last known
# good and bad revisions. Keep going until it says something like
# "First known bad revision is XYZ."
```

If you know a more recent revision that does work you can pass the hash instead of the v1.9.3 tag, then you'll have  fewer steps to test. I'd be very surprised if this specific PR broke ranger this way.

**Comment by manujchandra** [comment]
Hi @toonn 

Ranger is working in Bash. But not in Xonsh. I have not tested the setinregex yet.

I will do the bisect and will let you know.
